File: lib/src/main/java/com/tencent/cubershi/plugin_loader/blocs/CreateApplicationBloc.java
package com.tencent.cubershi.plugin_loader.blocs;

import android.app.Application;

import com.tencent.cubershi.plugin_loader.exceptions.CreateApplicationException;

/**
 * 初始化插件Application类
 *
 * @author cubershi
 */
public class CreateApplicationBloc {
    public static Application callPluginApplicationOnCreate(ClassLoader pluginClassLoader, String appClassName) throws CreateApplicationException {
        try {
            final Class<?> appClass = pluginClassLoader.loadClass(appClassName);
            final Application mockApplication = Application.class.cast(appClass.newInstance());
            mockApplication.onCreate();
            return mockApplication;
        } catch (Exception e) {
            throw new CreateApplicationException(e);
        }
    }
}

File: shadow-sdk/shadow-loader/src/main/kotlin/com/tencent/shadow/loader/blocs/CreateApplicationBloc.kt
package com.tencent.shadow.loader.blocs

import android.content.Context
import android.content.res.Resources
import com.tencent.shadow.loader.classloaders.PluginClassLoader
import com.tencent.shadow.loader.exceptions.CreateApplicationException
import com.tencent.shadow.loader.managers.ComponentManager
import com.tencent.shadow.loader.managers.PluginPackageManager
import com.tencent.shadow.runtime.ShadowApplication

/**
 * 初始化插件Application类
 *
 * @author cubershi
 */
object CreateApplicationBloc {
    @Throws(CreateApplicationException::class)
    fun createShadowApplication(
            pluginClassLoader: PluginClassLoader,
            appClassName: String,
            pluginPackageManager: PluginPackageManager,
            resources: Resources,
            hostAppContext: Context,
            componentManager: ComponentManager,
            receivers: Map<String, String>
    ): ShadowApplication {
        try {
            val appClass = pluginClassLoader.loadClass(appClassName)
            val shadowApplication = ShadowApplication::class.java.cast(appClass.newInstance())
            shadowApplication.setPluginResources(resources)
            shadowApplication.setPluginClassLoader(pluginClassLoader)
            shadowApplication.setPluginComponentLauncher(componentManager)
            shadowApplication.setHostApplicationContextAsBase(hostAppContext)
            shadowApplication.setReceivers(receivers)
            shadowApplication.setPluginPackageManager(pluginPackageManager)
            shadowApplication.setLibrarySearchPath(pluginClassLoader.getLibrarySearchPath())
            shadowApplication.setPluginPartKey(pluginPackageManager.pluginInfo.partKey)
            return shadowApplication
        } catch (e: Exception) {
            throw CreateApplicationException(e)
        }

    }
}
